d2core/d2asset/asset_manager.go

```diff
@@ -1,23 +1,228 @@
+	}
+
+	if err := singleton.Bind(fontManager); err != nil {
+		return err
+	}
+
+	if err := term.BindAction("assetspam", "display verbose asset manager logs", func(verbose bool) {
+		if verbose {
+			term.OutputInfof("asset manager verbose logging enabled")
+		} else {
+			term.OutputInfof("asset manager verbose logging disabled")
+		}
+
+		archiveManager.GetCache().SetVerbose(verbose)
+		archivedFileManager.GetCache().SetVerbose(verbose)
+		paletteManager.GetCache().SetVerbose(verbose)
+		am.paletteTransformManager.cache.SetVerbose(verbose)
+		animationManager.cache.SetVerbose(verbose)
+	}); err != nil {
+		return err
+	}
+
+	if err := term.BindAction("assetstat", "display asset manager cache statistics", func() {
+		var cacheStatistics = func(c d2interface.Cache) float64 {
+			const percent = 100.0
+			return float64(c.GetWeight()) / float64(c.GetBudget()) * percent
+		}
+
+		term.OutputInfof("archive cache: %f", cacheStatistics(archiveManager.GetCache()))
+		term.OutputInfof("file cache: %f", cacheStatistics(archivedFileManager.GetCache()))
+		term.OutputInfof("palette cache: %f", cacheStatistics(paletteManager.GetCache()))
+		term.OutputInfof("palette transform cache: %f", cacheStatistics(am.paletteTransformManager.
+			cache))
+		term.OutputInfof("animation cache: %f", cacheStatistics(animationManager.cache))
+		term.OutputInfof("font cache: %f", cacheStatistics(fontManager.GetCache()))
+	}); err != nil {
+		return err
+	}
+
+	if err := term.BindAction("assetclear", "clear asset manager cache", func() {
+		archiveManager.ClearCache()
+		archivedFileManager.GetCache().Clear()
+		paletteManager.ClearCache()
+		am.paletteTransformManager.cache.Clear()
+		animationManager.ClearCache()
+		fontManager.ClearCache()
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Bind a subordinate asset manager
+func (am *assetManager) Bind(manager d2interface.AssetManagerSubordinate) error {
+	switch _ := manager.(type) {
+	case d2interface.ArchiveManager:
+		if err := manager.Bind(am); err != nil {
+			return nil
+		}
+
+		am.archiveManager = manager.(d2interface.ArchiveManager)
+
+	case d2interface.ArchivedFileManager:
+		if err := manager.Bind(am); err != nil {
+			return nil
+		}
+
+		am.archivedFileManager = manager.(d2interface.ArchivedFileManager)
+
+	case d2interface.ArchivedPaletteManager:
+		if err := manager.Bind(am); err != nil {
+			return nil
+		}
+
+		am.paletteManager = manager.(d2interface.ArchivedPaletteManager)
+
+	case *paletteTransformManager:
+		if err := manager.Bind(am); err == nil {
+			am.paletteTransformManager = manager.(*paletteTransformManager)
+		}
+	case d2interface.ArchivedAnimationManager:
+		if err := manager.Bind(am); err != nil {
+			return err
+		}
+
+		am.animationManager = manager.(d2interface.ArchivedAnimationManager)
+
+	case d2interface.ArchivedFontManager:
+		if err := manager.Bind(am); err != nil {
+			return err
+		}
+
+		am.fontManager = manager.(d2interface.ArchivedFontManager)
+	}
+
+	return nil
```

Comment:
Panic on the default case

Reply:
this gets bubbled up to `app.Initialize`, which gets bubbled up to `app.Run`, which throws a fatal error

Reply:
nevermind, I see what you're talking about. adding it.